modules/apps/transfer/keeper/relay.go

```diff
@@ -310,6 +315,38 @@ func (k Keeper) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, data t
 	return nil
 }
 
+func (k Keeper) SetDenomMetaData(ctx sdk.Context, denomTrace types.DenomTrace) {
+	metadata := banktypes.Metadata{
+		Description: getMetaDataDescription(denomTrace),
+		DenomUnits: []*banktypes.DenomUnit{
+			{
+				Denom:    denomTrace.BaseDenom,
+				Exponent: 0,
+			},
+		},
+		// Setting base as IBChash Denom as SetDenomMetaData uses Base as storeKey
+		// and the bank keeper will only have the IBCHash to get the denom metadata
+		Base:    denomTrace.IBCDenom(),
+		Display: denomTrace.GetFullDenomPath(),
+		Name:    getMetaDataName(denomTrace),
+		Symbol:  getMetaDataSymbol(denomTrace),
```

Comment:
Can we have an example print out of how this might look? Would be great if we could get product input on the desired formatting? cc @womensrights @adiraviraj @tmsdkeys

Reply:
For context, this pr would add metadata about IBC tokens into the state machine. This means when clients query a bank balance, instead of only seeing `ibc/{hash}` they could also provide a flag to obtain back the metadata information stored here, such as the base denomination or the full history of this tokens hops (via channel/portIDs)

Reply:
I added a unit test for the new migration. The example metadata is available [here](https://github.com/0xmuralik/ibc-go/blob/dafd5f1cbbfaf77b3326ba7bdabffd480ccb80e7/modules/apps/transfer/keeper/migrations_test.go#L137).

Reply:
I've known people asking for this on Discord, to have the metadata about IBC tokens as you would regular sdk tokens so think it's a good idea. Wrt the formatting, looks okay for me. The `Base` field should normally equal `DenomUnits.Denom` with `Exponent` according to the bank module's metadata, but the reasoning is provided why it needs to deviate. 
Rest looks okay for me. Maybe @womensrights has insights from user research?

Reply:
I think this is a great first step, but also adding in a human readable name to the current output would precisely cover what the metadata is used for. Front end developers are currently maintaining their own registries of assets to be displayed for their application. An example of such a registry is [for Injective](https://github.com/InjectiveLabs/utils-go/blob/master/pkg/token/token_meta.json#L25-L36). Front ends need:

- token symbol, this is covered
- the location of the decimal place for the denom, I think this is covered from the exponent, from my understanding, e.g. the precision of ATOM is 6 dp, ATOM exponent is 6, uATOM is 0. In the example output given its not clear though as you have used an exponent of 0 for ubar and foo, foo should have exponent 6 assuming it is an sdk coin? 
- human readable name, having the path is good, but then you would need to check which chain the token came from additionally to make this human readable

So I think this is already an improvement, just if there was also a way to add an additional human readable name somehow then this would be great but I'm not sure this is easily done. Also as a sidenote, I know there is a constraint on the metadata from the sdk types so this is not easy to change, but for the ibc denoms, you don't get any new information from looking at the `Display` compared to `Name` so in the long term I would question the need to have both

Reply:
Thank you for the input, @womensrights!

- Regarding the exponent, if I am not mistaken I think at the moment it will be difficult to set it to the correct value, since this metadata is set on the receiving chain and the [`FungibleTokenPacketData` does not contain that information at the moment](https://github.com/cosmos/ibc-go/blob/cf519bc4d4285aa0b69511ba08d69b7456744f5d/proto/ibc/applications/transfer/v2/packet.proto#L10). However, [with channel upgradability it would be possible to extend the packet data and add extra information](https://github.com/cosmos/ibc/discussions/719).
- Regarding the lack of chain ID. Maybe the sending chain's chain ID could be added in the memo field? Not that I particularly want to continue adding more information in the memo file :), but maybe it's a solution for now. Your comment `if there was also a way to add an additional human readable name somehow` refers to adding the chain ID, right?

Reply:
For the exponent, as you linked the `FungibleTokenPacketData` there isn't explicitly an exponent field but where does the logic to know the decimal place for the amount in the packet for a specific denom come from in this case? 

The chainID would also not completely work as sometimes the chainID is not the same as what is considered the human branded name for the chain or you would need to remove the additional numbers. e.g. I think bandchain has chainID "laozi-mainnet" and the human readable name would be Bandchain

Reply:
> For the exponent, as you linked the FungibleTokenPacketData there isn't explicitly an exponent field but where does the logic to know the decimal place for the amount in the packet for a specific denom come from in this case?

The amount comes from the [sdk Coin that is used in `MsgTransfer`](https://github.com/cosmos/ibc-go/blob/main/proto/ibc/applications/transfer/v1/tx.proto#L38). [Whatever that amount is, we put it in the `FungibleTokenPacketData`](https://github.com/cosmos/ibc-go/blob/main/modules/apps/transfer/keeper/relay.go#LL132C1-L132C1), but we don't have any information about the exponent.

> The chainID would also not completely work as sometimes the chainID is not the same as what is considered the human branded name for the chain or you would need to remove the additional numbers. e.g. I think bandchain has chainID "laozi-mainnet" and the human readable name would be Bandchain

True, good point! Then I guess we cannot do anything about this at the moment?

Reply:
ok I see, I guess [this adr](https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-024-coin-metadata.md) has not yet been implemented and right now its assumed that sdk coins all have 6 decimal places.

Yes I don't think there can be anything done about the chainIDs